controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java

```diff
@@ -169,12 +168,21 @@ private static AttributeMarshaller credentialReferenceAttributeMarshaller() {
             @Override
             public void marshallAsElement(AttributeDefinition attribute, ModelNode credentialReferenceModelNode, boolean marshallDefault, XMLStreamWriter writer) throws XMLStreamException {
                 writer.writeStartElement(attribute.getXmlName());
-                if (credentialReferenceModelNode.hasDefined(clearTextAttribute.getName())) {
-                    clearTextAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
+                ModelNode crNode = null;
+                if (credentialReferenceModelNode.hasDefined(attribute.getName())) {
+                    crNode = credentialReferenceModelNode.get(attribute.getName());
                 } else {
-                    credentialStoreAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
-                    credentialAliasAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
-                    credentialTypeAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
+                    crNode = credentialReferenceModelNode;
```

Comment:
The existence of this 'else' block and the name of this 'credentialReferenceModelNode' param tells me this has gone wrong. The ModelNode param is meant to be a ModelNode that is a parent, with this attribute as a child. Generally that means it is a ModelNode that represents the Resource, with children for all attributes in that resource. The param is not meant to be the attribute value itself. If it ever is I fear something has gone wrong.

Have a look at DefaultAttributeMarshaller for the standard kind of implementation. A check of isMarshallable() with no xml writing at all if that returns false, and then the attribute data is extracted from the model as you do here in L173.